src/WorldIDIdentityManagerImplV3.sol

```diff
@@ -0,0 +1,233 @@
+        uint256[2] commitmentPok;
+
+        /// @notice KZG commitment for the polynomial extrapolated from the identities.
+        uint128[3] kzgCommitment;
+
+        /// @notice KZG proof associated with the commitment.
+        uint128[3] kzgProof;
+
+        /// @notice Expected evaluation of the polynomial at a certain point equal to kzgChallenge.
+        uint256 expectedEvaluation;
+
+        /// @notice The value for the root of the tree before the `identityCommitments` have been inserted.
+        ///         Must be an element of the field `Kr`. (already in reduced form)
+        uint256 preRoot;
+
+        /// @notice The root obtained after inserting all of `identityCommitments` into the tree described
+        ///         by `preRoot`. Must be an element of the field `Kr`. (already in reduced form)
+        uint256 postRoot;
+
+        /// @notice Challenge value used in the KZG proof to evaluate the polynomial.
+        uint256 kzgChallenge;
+
+        /// @notice Hash of the input data calculated as described in the comment
+        ///         to `calculateIdentityRegistrationInputHash()`.
+        bytes32 inputHash;
+
+        /// @notice Number of identities being registered in this batch.
+        uint32 batchSize;
+
+        /// @notice The position in the tree at which the insertions were made.
+        uint32 startIndex;
+    }
+
+    /// @notice Registers identities into the WorldID system following the EIP-4844 scheme.
+    /// @dev Can only be called by the identity operator.
+    /// @dev Registration is performed off-chain and verified on-chain via insertion proof
+    ///      and KZG proof. This saves gas and time over inserting identities one at a time.
+    ///
+    /// @param params parameters for the process defined by the `RegisterIdentities4844Params` structure.
+    /// @custom:reverts Unauthorized If the message sender is not authorised to add identities.
+    /// @custom:reverts NotLatestRoot If the provided `params.preRoot` is not the latest root.
+    /// @custom:reverts ProofValidationFailure If `params.insertionProof` cannot be verified using the
+    ///                 provided inputs.
+    /// @custom:reverts VerifierLookupTable.NoSuchVerifier If the batch sizes doesn't match a known
+    ///                 verifier.
+    function registerIdentities(
+        RegisterIdentities4844Params calldata params
+    ) public virtual onlyProxy onlyInitialized onlyIdentityOperator {
+        if (params.preRoot != _latestRoot) {
+            revert NotLatestRoot(params.preRoot, _latestRoot);
+        }
+
+        // No matter what, the inputs can result in a hash that is not an element of the scalar
+        // field in which we're operating. We reduce it into the field before handing it to the
+        // verifier. All other elements that are passed as calldata are reduced in the circuit.
+        uint256 reducedElement = uint256(params.inputHash) % SNARK_SCALAR_FIELD;
```

Comment:
How could this ever be not-reduced? It's an output of poseidon, therefore always a field element. Just use plain input hash